ipq807x: change 3 devices to use the standard eMMC sysupgrade code

Lanchon opened the pull request:
Settings were lost because target-specific `mmc_do_upgrade` does not align backup data to 64 KiB boundaries.

This changes the device to use the generic `emmc_do_upgrade` and `emmc_copy_config`.

---

i don't have this device, but it seems all devices using `mmc_do_upgrade` should be affected.

@pwned-pixel did the testing and reportedly this change fixes sysupgrade: settings are preserved while random files in the overlay (files not explicitly backed up) are deleted.

---

@dangowrt @robimarko

thank you!

Lanchon committed:
ipq807x: nbg7815: use the standard eMMC sysupgrade code

Note that the old ad-hoc method did not explicitly align backup data
to 64 KiB boundaries.

Signed-off-by: Rodrigo Balerdi <lanchon@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16505
Signed-off-by: Robert Marko <robimarko@gmail.com>

Lanchon commented:
after this change only two `ipq807x` devices remain with the supposedly faulty `mmc_do_upgrade`:
- prpl haze
- qnap 301w

i could move them to `emmc_do_upgrade` / `emmc_copy_config` but i can't test them.

robimarko commented:
Qnap 301W is not broken, I have it and it preserves settings just fine.

That being said, If it can be moved to a generic sysupgrade EMMC function instead of the custom one I am all for it (Though I am not a fan of F2FS)

Lanchon commented:
@robimarko 

> Qnap 301W is not broken

ok. but how is it not broken? AFAICT, fstools creates the loop device rounding up the start address to 64 KiB, and `mmc_do_upgrade` does not:
https://github.com/openwrt/openwrt/blob/38bb47c36c473a33b054a76dcd52827ae15fb610/target/linux/qualcommax/ipq807x/base-files/lib/upgrade/mmc.sh#L53-L59

so unless the rootfs itself is somehow padded to 64 KiB for this device, i don't understand how this could be working.

Lanchon commented:
> (Though I am not a fan of F2FS)

me neither, but the powers that be have chosen F2FS for overlay.

(i don't think that replacing the generic emmc upgrade code with a custom one for each device is a rational way of dealing with this.)

btw, here is a PR related to this: https://github.com/openwrt/fstools/pull/9

Lanchon commented:
@robimarko 

> Qnap 301W is not broken, I have it and it preserves settings just fine.

after looking into it a bit more, i'm willing to bet you are wrong.

the device has a `rootfs_data` partition:
```
  9     0x00322022      0x0073d821      "rootfs_data"
        attrs:  0x0000000000000000
        type:   ab1760da-a8bb-4d6f-98d2-9ad3ab9009cd
        guid:   d1402149-3e8e-0709-99d9-170bfab2f66c
```

but `mmc_do_upgrade` does not support that. instead it writes to rootfs loop, in the wrong location, but it doesn't matter because openwrt mounts `rootfs_data` anyway after boot. this is the same issue the spectrum had.

so these are symptoms of sysupgrade failing:
- write a random file in /, sysupgrade, and verify that the file DIDN'T go away
- sysupgrade wiping settings, and verify that settings are conserved

this happens because `rootfs_data` isn't touched at all during sysupgrade.

Lanchon commented:
@robimarko 

given that i understood the why sysupgrade appeared to work until now, i added a commit here fixing the problem.

since you own the device, please verify that there is a discreet `rootfs_data` data partition and that sysupgrade was not touching it before this change.

thanks!

Lanchon commented:
ok, and i pushed a commit for the haze, the last device using mmc.sh, given there is evidence that this last one does *not* use a discreet rootfs_data: https://github.com/openwrt/openwrt/commit/ba415af5706f0369e4d299e4dfe50541a4a8056b

so mmc.sh is still broken but now it's unused, AFAICT.

i want to delete it but i don't know who is referencing it at runtime!

kirdesde commented:
I already wrote in https://github.com/openwrt/openwrt/issues/16452#issuecomment-2366838143 that sysupgrade + preserving config is working fine (at least for 301w and the nbg7815)

The only thing that might not work is sysupgrade and **not** preserving config since mmc_do_flash is always calling cp backup config.

https://github.com/openwrt/openwrt/blob/676a2ab03b37895ae47f7730cb9abcee38ebe774/target/linux/qualcommax/ipq807x/base-files/lib/upgrade/mmc.sh#L74

BTW IMHO your change for the  prpl,haze would break the sysupgrade since this device doesn't have a mmc rootfs_data partition and therefore needs the roofs_data loop.

Lanchon commented:
> I already wrote in [#16452 (comment)](https://github.com/openwrt/openwrt/issues/16452#issuecomment-2366838143) that sysupgrade + preserving config is working fine (at least for 301w and the nbg7815)
> 
> The only thing that might not work is sysupgrade and **not** preserving config since mmc_do_flash is always calling cp backup config.
> 
> https://github.com/openwrt/openwrt/blob/676a2ab03b37895ae47f7730cb9abcee38ebe774/target/linux/qualcommax/ipq807x/base-files/lib/upgrade/mmc.sh#L74

i already wrote i detailed reply to your comment [here](https://github.com/openwrt/openwrt/issues/16452#issuecomment-2367019492), which you seem to have completely ignored. in it i explain how sysupgrade is supposed to work.

i also explained why "sysupgrade + preserving config is working fine" is not proof that sysupgrade is working.

also, "The only thing that might not work" is very probably *not* the only thing not working. again, see my reply for more info.

> BTW IMHO your change for the prpl,haze would break the sysupgrade since this device doesn't have a mmc rootfs_data partition and therefore needs the roofs_data loop.

with all due respect, you don't seem to undertand how sysupgrade works. no, you *do not* need a loop device *during sysugrade*. you do need it after booting though, which is the responsibility of fstools to create.

and note that any loop device you might create during sysupgrade will cease to exist on reboot, as loop devices exist in kernel ram and are not persistent.

again, refer to my previous reply to you, especially the section mentioning the WHW03 v1.

Lanchon commented:
@dangowrt was that an intention to merge?

sppmasterspp commented:
> @dangowrt was that an intention to merge?

I have 301w.
Can I help with anything resolving this?

Lanchon commented:
@sppmasterspp 

> I have 301w.
> Can I help with anything resolving this?

thank you! yes, please.

you can confirm that the current 301w sysupgrade code **does not work** by running these tests with the official openwrt firmware:

1. create any random file in the root `/` directory of the router, then do a sysupgrade **preserving** settings. next check if the the random file you created still exists or is gone:
   - if the file still exists, sysupgrade is broken.

2. make some config changes, then do a sysupgrade **resetting** settings. next check if router settings were actually reset:
   - if settings were not reset, sysupgrade is broken.

thanks again.

sppmasterspp commented:
> you can confirm that the current 301w sysupgrade code **does not work** by running these tests with the current official openwrt firmware:
> 
> 1. create any random file in the root `/` directory of the router, then do a sysupgrade **preserving** settings. next check if the the random file you created still exists or is gone:
>    
>    * if the file still exists, sysupgrade is broken.

For now I've been able to test only the first one because I had a newly compiled build form 2 hours ago. 
A big notice here is that I use NSS build (not officially supported by OpenWrt upstream). Build is based on @qosmio NSS-wifi branch.
I created a file `PR16505` in root / with content 301w but after sysupgrade preserving the settings the file is gone after reboot.

Lanchon commented:
@sppmasterspp

please only test running official openwrt both **before and after** the test.

forks are not supported.

Lanchon commented:
@robimarko 

i haven't heard back from you regarding:
https://github.com/openwrt/openwrt/pull/16505#issuecomment-2375585324

qnap 301W sysupgrade in the current official builds is very probably broken and should be tested like this:
https://github.com/openwrt/openwrt/pull/16505#issuecomment-2382645037

there is a commit here to fix it.

thanks!

ynezz commented:
1. I did flashed in U-Boot latest sysupgrade image `r27934-ff18576f84`
2. I've set `uci set system.@system[0].hostname='OpenWrtuned' ; uci commit system` to see, that the config can be preserved after boot
3. I've added cookie `echo foo > /bar`
4. I did `sysupgrade https://downloads.openwrt.org/snapshots/targets/qualcommax/ipq807x/openwrt-qualcommax-ipq807x-prpl_haze-squashfs-sysupgrade.bin` 

```shell
root@OpenWrtuned:~# ubus call system board
{
        "kernel": "6.6.58",
        "hostname": "OpenWrtuned",
        "system": "ARMv8 Processor rev 4",
        "model": "prpl Foundation Haze",
        "board_name": "prpl,haze",
        "rootfs_type": "squashfs",
        "release": {
                "distribution": "OpenWrt",
                "version": "SNAPSHOT",
                "description": "OpenWrt SNAPSHOT",
                "revision": "r27934-ff18576f84",
                "target": "qualcommax/ipq807x",
                "builddate": "1730108090"
        }
}
```

> write a random file in /, sysupgrade, and verify that the file DIDN'T go away

The `/bar` cookie went away :white_check_mark: This is expected behavior as its not defined in `/lib/upgrade/keep.d` ?

> sysupgrade wiping settings, and verify that settings are conserved

5. `sysupgrade -n https://downloads.openwrt.org/snapshots/target
s/qualcommax/ipq807x/openwrt-qualcommax-ipq807x-prpl_haze-squashfs-sysupgrade.bin`

````shell
root@OpenWrt:~# ubus call system board
{
        "kernel": "6.6.58",
        "hostname": "OpenWrt",
        "system": "ARMv8 Processor rev 4",
        "model": "prpl Foundation Haze",
        "board_name": "prpl,haze",
        "rootfs_type": "squashfs",
        "release": {
                "distribution": "OpenWrt",
                "version": "SNAPSHOT",
                "description": "OpenWrt SNAPSHOT",
                "revision": "r27934-ff18576f84",
                "target": "qualcommax/ipq807x",
                "builddate": "1730108090"
        }
}
```

sysupgrade wiped settings :white_check_mark:

Lanchon commented:
@ynezz 

> test on prpl,haze

thanks! unfortunately these tests were intended for qnap 301w, which uses a discrete rootfs_data partition for the overlay. these tests never applied to haze.

haze is different and uses the trailing part of rootfs for the overlay. on haze the expectation was that sysupgrade would possibly always clear settings even if you wanted to keep them, because the update code does not align the loop device to 64KB boundaries. however your test shows that settings are kept. i suspect this means that rootfs is generated in such a way that causes it to always be a multiple of 64KB in size? idk

whatever the reason, even if it worked, this is brittle. and haze should be migrated to the standard way of doing things anyway, which [this commit](https://github.com/openwrt/openwrt/pull/16505/commits/bb29b6fc84ffd8f0c4fda622b0f34e30f99dbcca) does, even if it is currently working. my point is that i'm pretty sure qnap 301w **does not work**, and the others need to be migrated to the standard code even if they work. @robimarko has a qnap 301w but unfortunately he hasn't tested it.

sppmasterspp commented:
> @ynezz
> 
> > test on prpl,haze
> 
> thanks! unfortunately these tests were intended for qnap 301w, which uses a discrete rootfs_data partition for the overlay. these tests never applied to haze.
> 
> haze is different and uses the trailing part of rootfs for the overlay. on haze the expectation was that sysupgrade would possibly always clear settings even if you wanted to keep them, because the update code does not align the loop device to 64KB boundaries. however your test shows that settings are kept. i suspect this means that rootfs is generated in such a way that causes it to always be a multiple of 64KB in size? idk
> 
> whatever the reason, even if it worked, this is brittle. and haze should be migrated to the standard way of doing things anyway, which [this commit](https://github.com/openwrt/openwrt/pull/16505/commits/bb29b6fc84ffd8f0c4fda622b0f34e30f99dbcca) does, even if it is currently working. my point is that i'm pretty sure qnap 301w **does not work**, and the others need to be migrated to the standard code even if they work. @robimarko has a qnap 301w but unfortunately he hasn't tested it.

I really want to help with this if I can at least when I have spare time and the network is not busy. Is there a simple way to test this?

ynezz commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -308,7 +308,7 @@ define Device/prpl_haze
 	SOC := ipq8072
 	DEVICE_PACKAGES := ath11k-firmware-qcn9074 ipq-wifi-prpl_haze kmod-ath11k-pci \
-		mkf2fs f2fsck kmod-fs-f2fs kmod-leds-lp5562
+		kmod-fs-f2fs f2fs-tools kmod-leds-lp5562
```
I miss description/reasoning about  this change in the commit. If its unrelated, move it into separate commit with proper description.

Lanchon replied:
description/reasoning: mkf2fs, f2fsck are part of metapackage f2fs-tools. f2fs-tools will be extended/adapted as f2fs devels see fit, instead of stagnating in the past. including the metapackage is the standard practice on other devices in openwrt, AFAICT.

> If its unrelated, move it into separate commit with proper description

i fixed [the device i own](https://github.com/openwrt/openwrt/commit/3b221ba5afe42bc32e878299d7d2275702536684). i don't have this device, i just wanted to help when i identified issues in other devices. you are more than welcome to take over these changes if you favor particular styles. in any case, thanks for the comment.

ynezz commented on target/linux/qualcommax/ipq807x/base-files/lib/upgrade/mmc.sh:
```diff
@@ -1,83 +0,0 @@
-#
-# Copyright (C) 2016 lede-project.org
```
This should go into separate commit as well.

Lanchon replied:
this change was lumped in this commit because it was the last device using this non-standard code, assuming the previous 2 commits were merged.

ynezz commented:
> whatever the reason, even if it worked, this is brittle. and haze should be migrated to the standard way of doing things anyway

I agree and I actually like it, but I'm only able to test this on Haze, but if it works, I'm going to merge only Haze relevant commits, unless you're willing to split this into Haze only pull request. I'm not going to merge the other bits as I don't want to cause any regressions as I wont be able to handle them in timely manner.

Lanchon commented:
@ynezz 

i thought these 3 commits would be an easy merge, especially after 3b221ba5afe42bc32e878299d7d2275702536684 was merged. i was wrong.

please go ahead and do your fix for Haze. one device fixed is better than none.

thanks for the help!

Lanchon commented:
@ynezz

a bit of history:

this PR request was about the first commit only, affecting only nbg7815, as you can see in the original PR message. later i saw that there were 2 other similar devices with issued and added commit 2 and 3.

but nbg7815 was already tested and the issue confirmed when i created the fix and this PR (again see OP, it details who tested). 

so you can definitely merge the first commit, and with you own fix to haze, only qnap 301w would remain broken.

sppmasterspp commented:
> @sppmasterspp
> 
> > I have 301w.
> > Can I help with anything resolving this?
> 
> thank you! yes, please.
> 
> you can confirm that the current 301w sysupgrade code **does not work** by running these tests with the current official openwrt firmware:
> 
>     1. create any random file in the root `/` directory of the router, then do a sysupgrade **preserving** settings. next check if the the random file you created still exists or is gone:
>        
>        * if the file still exists, sysupgrade is broken.
> 
>     2. make some config changes, then do a sysupgrade **resetting** settings. next check if router settings were actually reset:
>        
>        * if settings were not reset, sysupgrade is broken.
> 
> 
> thanks again.
@robimarko is probably very busy.
Tests on QNAP 301w.
Previously I've tested this with my custom forked NSS build. With it none of the two (point 1 and 2) can be observed.
I've just tested all of this with latest official Vanilla OpenWrt build.
First I flashed the official OpenWrt twice not keeping the settings with reboots of course.
Then I created a file in root / with content 301w. 
Next I followed point 1 - I did a sysupgrade keeping the settings - after reboot the file I created in root / wasn't there.
Point 2 - I made changes to some settings and saved them. I rebooted manually and checked that the settings were preserved - they were preserved.
I did a sysupgrade with resetting the setting. After reboot all the settings were reset to default.
I don't know if all of this helps. I may try another tests if necessary.

Lanchon commented:
@sppmasterspp 

> I don't know if all of this helps.

it helps a lot, thank you for testing. i was wrong!

i don't have the 301w, and when researching it, i came across what i now think might have been false info. as stated [here](https://github.com/openwrt/openwrt/pull/16505#issuecomment-2375585324), i was under the assumption that the 301w had a discrete `rootfs_data` partition. i don't know exactly where i got that snippet of a supposed dump of the 301w partition table, most likely the 301w forum.

now i doubt that that info is true. if the 301w *had* a rootfs_data, it would be failing in the same way as the spectrum sax1v1k did before i fixed it. your tests clearly show sysupgrade works, and make me think there is no discreet rootfs_data.

i sincerely apologize for the mistake.

under the new assumption, 301w would still benefit from a change to the new way of doing things. **but my proposed commit is wrong and will break sysupgrade if merged.**

@sppmasterspp could you please install package `gdisk` and use it to print the partition table of your device? you could do this on any build you want. this would confirm that GPT snippet i found earlier was red hearing, and i could at least provide a equivalent change for what @ynezz is requesting for haze [here](https://github.com/openwrt/openwrt/pull/16505#issuecomment-2444335271) (along with haze).

sppmasterspp commented:
> @sppmasterspp could you please install package `gdisk` and use it to print the partition table of your device? you could do this on any build you want. this would confirm that GPT snippet i found earlier was red hearing, and i could at least provide a equivalent change for what @ynezz is requesting for haze [here](https://github.com/openwrt/openwrt/pull/16505#issuecomment-2444335271) (along with haze).


Is this info enough?
```
root@QNAP:~# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00050000 00010000 "0:sbl1"
mtd1: 00010000 00010000 "0:mibib"
mtd2: 00180000 00010000 "0:qsee"
mtd3: 00010000 00010000 "0:devcfg"
mtd4: 00010000 00010000 "0:apdp"
mtd5: 00040000 00010000 "0:rpm"
mtd6: 00010000 00010000 "0:cdt"
mtd7: 00020000 00010000 "0:appsblenv"
mtd8: 00100000 00010000 "0:appsbl"
mtd9: 00040000 00010000 "0:art"
mtd10: 00080000 00010000 "0:ethphyfw1"
mtd11: 00080000 00010000 "0:ethphyfw2"
mtd12: 00350000 00010000 "reserved"
root@QNAP:~# gdisk /dev/mmcblk0
GPT fdisk (gdisk) version 1.0.10

Partition table scan:
  MBR: protective
  BSD: not present
  APM: not present
  GPT: present

Found valid GPT with protective MBR; using GPT.

Command (? for help): p
Disk /dev/mmcblk0: 7634944 sectors, 3.6 GiB
Sector size (logical/physical): 512/512 bytes
Disk identifier (GUID): xxxxxxxx-xxxx-xxxx-xxxxx-xxxxxxxxxx
Partition table holds up to 12 entries
Main partition table begins at sector 2 and ends at sector 4
First usable sector is 34, last usable sector is 7634910
Partitions will be aligned on 2-sector boundaries
Total free space is 42941 sectors (21.0 MiB)

Number  Start (sector)    End (sector)  Size       Code  Name
   1              34           32801   16.0 MiB    FFFF  0:HLOS
   2           32802           65569   16.0 MiB    FFFF  0:HLOS_1
   3           65570           98337   16.0 MiB    FFFF  0:HLOS_2
   4           98338         1146913   512.0 MiB   FFFF  rootfs
   5         1146914         2195489   512.0 MiB   FFFF  rootfs_1
   6         2195490         3244065   512.0 MiB   FFFF  rootfs_2
   7         3244066         3252257   4.0 MiB     FFFF  0:WIFIFW
   8         3252258         3285025   16.0 MiB    8301  reserved
   9         3285026         7591969   2.1 GiB     FFFF  rootfs_data

Command (? for help):
```

Lanchon commented:
@sppmasterspp 

> I've added this info to the QNAP 301w wiki.

thanks, that's great!

> Is this info enough?

could you do one more thing? please paste the output of the `mount` command.

> I don't know what partition 9 is for. Reserved for something?
There is a question for all the devices that have eMMC flash memory about those reserved partitions if this 2.1GiB can be used for anything useful at all anytime.

partition 9 has label `rootfs_data`. openwrt uses magic label names during boot to mount stuff like ubi partitions, the overlay fs, etc. `rootfs_data` is one such magic label, and i would have thought that that 2.1 GB partition would have been mounted as overlay (just like it happens on spectrum sax1v1k). and then... sysupgrade would have been broken.

but given that sysupgrade works, the overlay might be mounted instead as a loop device over one of the `rootfs` partitions for unknown reasons. this is what the output of `mount` will elucidate. this second case would provide near 500 MB for the overlay, instead of the possible 2.1 GB of the first case. this can be checked in LuCI in System/Software/Disk space.

in the spectrum sax1v1k, the `rootfs_data` is mounted as overlay as as expected, and then sysupgrade was broken (as expected) and required fixing. the difference is with the 301w is that your device has both bare nand flash *and* an emmc, while the spectrum only has an emmc.

the code that mounts the overlay is in repo fstools, and i'm [proposing changes](https://github.com/openwrt/fstools/pull/9) to make the magic labels configurable, but so far the code is unreviewed.

sppmasterspp commented:
> could you do one more thing? please paste the output of the `mount` command.

```
root@QNAP:~# mount
/dev/root on /rom type squashfs (ro,relatime,errors=continue)
proc on /proc type proc (rw,nosuid,nodev,noexec,noatime)
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,noatime)
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)
tmpfs on /tmp type tmpfs (rw,nosuid,nodev,noatime)
/dev/loop0 on /overlay type ext4 (rw,noatime)
overlayfs:/overlay on / type overlay (rw,noatime,lowerdir=/,upperdir=/overlay/upper,workdir=/overlay/work,xino=off)
tmpfs on /dev type tmpfs (rw,nosuid,noexec,noatime,size=512k,mode=755)
devpts on /dev/pts type devpts (rw,nosuid,noexec,noatime,mode=600,ptmxmode=000)
debugfs on /sys/kernel/debug type debugfs (rw,noatime)
bpffs on /sys/fs/bpf type bpf (rw,nosuid,nodev,noexec,noatime,mode=700)
pstore on /sys/fs/pstore type pstore (rw,noatime)
/dev/sda1 on /mnt/sda1 type f2fs (rw,lazytime,relatime,background_gc=on,nogc_merge,discard,discard_unit=block,user_xattr,inline_xattr,acl,inline_data,inline_dentry,flush_merge,barrier,extent_cache,mode=adaptive,active_logs=6,alloc_mode=default,checkpoint_merge,fsync_mode=posix,memory=normal,errors=continue)
root@QNAP:~#
```

Lanchon commented:
@sppmasterspp thank you for this info!

so yes, the partition is there, and i wasn't mistaken when i gathered that info from the forums:
```
   9         3285026         7591969   2.1 GiB     FFFF  rootfs_data
```

the surprising fact here is that openwrt is ignoring it. AFAIK, openwrt should use rootfs_data for overlay, but instead it is using a loop on the rootfs device, resulting in the < 500 MB storage:
```
/dev/loop0 on /overlay type ext4 (rw,noatime)
```

note that this is not what is happening in the similar spectrum sax1v1k: in that device, rootfs_data is used for overlay. this broke sysupgrade because the port was done based on QNAP 301w (the inherited sysupgrade code from 301w expected a loop device, not a discreet overlay partition).

a notable difference between these devices is that the spectrum only has emmc, while the 301w also has bare flash. maybe this dual storage is tripping a bug in openwrt fstools that makes it ignore rootfs_data?

in any case, this could be considered to be a lucky accident. IDK if the current port for 301w supports dual/triple firmware, but if it does, using a loop device allows for each firmware to have its separate configuration. and i think 500 MB is plenty of space for the overlay.

also note that during sysupgrade, the whole overlay is wiped and written down again (usually twice). which means that the kept data (does not include packages) must fit it RAM for the process to succeed. and also that the overlay is no place for valuable data. so in big devices i prefer a 500 MB overlay plus a large extra partition that survives sysupgrades, that those few who need it can use. (this is my own view, not necessarily shared by the project.)

a problem i see here is, if you use rootfs_data for extra storage, and this suspected bug is fixed somewhere down the line, your data will be silently wiped on a subsequent sysupgrade. you could preempt this situation by renaming the rootfs_data partition (gdisk) if you plan to use it. (or if [this PR](https://github.com/openwrt/fstools/pull/9) is ever merged, you could instead alter your uboot environment to make openwrt look for a partition that doesn't exist instead of rootfs_data.)

**WARNING: do not alter the GPT in any way (eg: renaming a partition) if your device has secure boot enabled! it is likely that the GPT is signed, and altering any part of it will lead to an irrecoverable brick.**

sppmasterspp commented:
> i prefer a 500 MB overlay plus a large extra partition that survives sysupgrades, that those few who need it can use. (this is my own view, not necessarily shared by the project.)

Thanks for your great work and useful suggestions for advanced usage. 
At this time I would really like to use the extra partition for saving additional data (like e.g. backups for adblock, statistics, SSH history, rclone config) that survives sysupgrades although I will need some more knowledge and caution so I can do this.
I really like it and I hope other developers look at this and we can get even more possibilities for advanced usage of this extra partition,

> maybe this dual storage is tripping a bug in openwrt fstools that makes it ignore rootfs_data?
and if it's a bug it will be fixed.
Probably there are other advanced users that will be interested in this too.

> WARNING: do not alter the GPT in any way (eg: renaming a partition) if your device has secure boot enabled! it is likely that the GPT is signed, and altering any part of it will lead to an irrecoverable brick.

I think the secure boot option is not used/enabled on 301w but maybe @kirdesde @robimarko can confirm this and make their comments/remarks on your proposals. Anyway I do not intend to do something risky with the partition layout in any way until more knowledgeable users confirm it is safe to do it.

robimarko commented:
My 301W does not have secure boot enabled

pwned-pixel commented:
> a notable difference between these devices is that the spectrum only has emmc, while the 301w also has bare flash. maybe this dual storage is tripping a bug in openwrt fstools that makes it ignore rootfs_data?

For the NBG7815 rootfs_data isn't used too.

But there is a bug within fstools regarding extroot/automount for the NBG7815. If blockd resp. block-mount is installed wifi does not come up on first boot. The fstab_init script is stalled like forever. This is caused (imo; I don't really have a clue) because block.c is searching only for mtd resp. ubi partitions. There are some workarounds and patches for the symptoms. But to me it looks like that rootfs_data is in the same boat here not getting included.

Maybe this observation helps to narrow this down.

> also note that during sysupgrade, the whole overlay is wiped and written down again (usually twice). which means that the kept data (does not include packages) must fit it RAM for the process to succeed.

This part scares me a bit. Because I've re-parted two of my devices >= 1GB for rootfs. Personally I don't care about data preservation at all. By default it should not hurt for NBG7815 because rootfs is limited to 2x60MB (haha).


Regarding the topic. I'm using this patch since 1 month for my builds on NBG7815 without any issue.

Lanchon commented:
@pwned-pixel

> Regarding the topic. I'm using this patch since 1 month for my builds on NBG7815 without any issue.

nonetheless, but the NBG7815 commit here shouldn't be cherrypicked and merged. although the code is correct, the commit message is not: it states that there is an issue in NBG7815 that i no longer think exists. the motivation for this commit should instead be changing to the new way of doing things.

@sppmasterspp 

> Is there any concern about wearing out the flash (eMMC in this case) with frequent writing to it?

i wouldn't have any concerns at all. eMMC does wear leveling, distributing writes over the entire flash area. in any kind of normal use, the flash will outlast the device. you can see this in phones for instance, which write to the flash heavily, and die for any reason but flash endurance. (early phones with issues eMMC, such as the samsung galaxy 2 i9100 from 2011, were cause by eMMC firmware bugs (eg: brickbug), not by endurance exhaustion.) just try not using virtual memory long term on the flash (but if you have to, reduce swappiness to zero or low values).

for write heavy applications, f2fs is probably easier on the flash than ext4 (with journaling, i.e. the default) and also faster, but is less well tested. for read heavy apps, i suppose both should perform equally. both fs should be acceptable for you use.

> I would really like to use the extra partition 9 on the eMMC for saving additional data

install the block-mount package, then reboot. in luci go to system/mount points. i'd suggest disabling all automatic mount functions and manually create a mount point for partition 9. btw, i always use the mount point "/extra" for this kind of extra storage. (and in some official openwrt MTD-based ports i also named the big unused partition as "extra".)

sppmasterspp commented:
> install the block-mount package, then reboot. in luci go to system/mount points. i'd suggest disabling all automatic mount functions and manually create a mount point for partition 9. btw, i always use the mount point "/extra" for this kind of extra storage. (and in some official openwrt MTD-based ports i also named the big unused partition as "extra".)

I've made those changes and mounted partition 9. There are some interesting stuff inside.
![image](https://github.com/user-attachments/assets/a6b37699-aaa5-4939-898e-a3314e3b1295)

![image](https://github.com/user-attachments/assets/9bf11894-2d06-4405-9a0a-02f27833bdde)
Seems that this partition was used by the QNAP stock firmware to write to and to save data to it.
Now the question is how that is or may be related to the questions discussed here and the fact the partition 9 is not used by OpenWrt.
I won't touch any files/directories on it just to keep the data safe and intact.
Maybe I can create a new folder (/root/extra) for my own usage and mount /extra folder only and save my OpenWrt data only in that folder.

Lanchon commented:
@sppmasterspp 

`df -h` will probably show that p9 is mostly empty. with stock firmware, p9 is probably wiped when the device is reset, meaning that to restore stock you just need to wipe it... but i could be wrong. if you don't plan to ever restore stock you can just wipe it. otherwise you can back it up.

if you want to back up the partition as a binary image, and the partition is mostly empty: mount it, run `fstrim -v /path/to/device/mount/point/` to wipe unused space, unmount it, `gzip -c /dev/<emmc-dev>p9 > /tmp/p9.img.gz` to back it up, and then use scp to pull the image backup to your PC.

or you can just backup the files in an archive, which should be enough: `tar --acls --xattrs --selinux -cpvzf /tmp/p9.tgz /path/to/device/mount/point`. options --acls --xattrs and --selinux may be too much for busybox tar, you can leave them out then. or if you want an uncompressed backup (faster) you can use: `tar --acls --xattrs --selinux -cpvf /tmp/p9.tar /path/to/device/mount/point`.

(if there is a lot of data that doesn't fit in /tmp, you can stream the compressed partition over ssh.)

after the backup, choose an fs and format the partition. f2fs should work better for write heavy loads. after mounting the new partition, make sure `fstrim -v` works on it. change mount options if not. no having trim/discard enabled can significantly impact performance and longevity of the emmc.

sppmasterspp commented:
> `df -h` will probably show that p9 is mostly empty. with stock firmware

Yes it's mostly empty (only 4% occupied) as can be seen on the picture I've put on my post too.
`
Filesystem                Size      Used Available Use% Mounted on
/dev/mmcblk0p9            2.0G     79.9M      1.8G   4% /mnt/extra

`
`
root@QNAP:~# fstrim -v /mnt/extra
/mnt/extra: 1.9 GiB (2016382976 bytes) trimmed
`
Then following you guide I've made both backups - as a binary image and as an archive.
`
root@QNAP:~# gzip -c /dev/mmcblk0p9 > /tmp/p9.img.gz
I don't use busybox tar so it was OK running below command.
root@QNAP:~# tar --acls --xattrs --selinux -cpvzf /tmp/p9.tgz /mnt/extra
tar: Removing leading `/' from member names
/mnt/extra/
/mnt/extra/proc/
/mnt/extra/media/
/mnt/extra/cfg/
/mnt/extra/cfg/default/
/mnt/extra/cfg/default/ini
/mnt/extra/profile.json
/mnt/extra/ini
.....
`


> after the backup, choose an fs and format the partition. f2fs should work better for write heavy loads

`
root@QNAP:~# mkfs.f2fs /dev/mmcblk0p9

    F2FS-tools: mkfs.f2fs Ver: 1.16.0 (2023-04-11)

Info: Disable heap-based policy
Info: Debug level = 0
Info: Trim is enabled
Info: Segments per section = 1
Info: Sections per zone = 1
Info: sector size = 512
Info: total sectors = 4306944 (2103 MB)
Info: zone aligned segment0 blkaddr: 512
Info: format version with
  "Linux version 6.6.58 (aarch64-openwrt-linux-musl-gcc (OpenWrt GCC 14.2.0 r27829-11038acff2) 14.2.0, GNU ld (GNU Binutils) 2.43.1) #"
Info: [/dev/mmcblk0p9] Discarding device
Info: This device doesn't support BLKSECDISCARD
Info: Discarded 2103 MB
Info: Overprovision ratio = 3.530%
Info: Overprovision segments = 37 (GC reserved = 35)
Info: format successful

root@QNAP:~# fstrim -v /mnt/extra
/mnt/extra: 0 B (0 bytes) trimmed

`
![image](https://github.com/user-attachments/assets/d5db64fc-ba5c-4e1f-9570-e0024e353af6)
I rebooted the router and started to configure some packages to use the extra space that should survive sysupgrades.
Finally I have it working
![image](https://github.com/user-attachments/assets/68297e10-96f5-4bcb-8636-79190d6d5c08)

I really appreciate your invaluable help so now this device is almost ready to fly to the Moon :).

Lanchon commented:
@robimarko 

> Qnap 301W is not broken, I have it and it preserves settings just fine.

ok so we now know that the 301w is doubly broken, lol:

- there is a bug in fstools that makes it ignore the `rootfs_data` partition: instead of mounting it as an overlay, it creates a loop device on `rootfs`.

- which means that the "bug" i was trying to correct (sysupgrade not using `rootfs_data`) was acting not as a bug but a workaround.

- but it should be noted that the minute this bug is fixed in fstools, sysupgrade on this device will break.

Lanchon committed:
ipq807x: prpl haze & qnap 301w: use the standard eMMC sysupgrade code

Note that the old ad-hoc method did not explicitly align backup data
to 64 KiB boundaries.

Also note that the qnap 301w has a 'rootfs_data' partition in the eMMC
that is being ignored by fstools during boot, presumably due to a bug.
This is why the partition is also ignored in the sysupgrade code and
there is no definition of CI_DATAPART="rootfs_data".

Signed-off-by: Rodrigo Balerdi <lanchon@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16505
Signed-off-by: Robert Marko <robimarko@gmail.com>

Lanchon committed:
ipq807x: delete unused legacy eMMC sysupgrade code

All ipq807x devices that were using the legacy 'mmc_do_upgrade' eMMC
sysupgrade code were ported to the replacement 'emmc_do_upgrade' code.

Signed-off-by: Rodrigo Balerdi <lanchon@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16505
Signed-off-by: Robert Marko <robimarko@gmail.com>

Lanchon committed:
ipq807x: order devices alphabetically

Alphabetically sort devices in platform.sh

Signed-off-by: Rodrigo Balerdi <lanchon@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16505
Signed-off-by: Robert Marko <robimarko@gmail.com>

Lanchon commented:
@robimarko 

> BTW, are NBG and Haze commits ready to be applied?

no. the code was correct, but the commit comments were not. there was also a request by @ynezz which was pending. i just fixed all this.

i also migrated 301w to the new emmc code, but left the workaround/bug untouched. so now commits for all 3 devices can be merged.

after those i also included 2 other cleanup commits, which i recommend merging (ONLY if the previous commits are accepted).

@ynezz 

> [the deletion of the unused old code] should go into separate commit as well.

thanks, done!

Lanchon commented:
@robimarko 

this is a side note:

> Well, it means we need to have a fix for 301w that will be applied at the same time as fstools is fixed.

actually i think this 301w "bug" of not using `rootfs_data` should never be fixed, the current behavior being more desirable, and i proposed an alternative solution [here](https://github.com/openwrt/fstools/issues/10).

Lanchon commented:
btw, one commit does not pass test formalities for missing extra commit text.

i think the text "order devices alphabetically" more than adequately explains [the commit](https://github.com/openwrt/openwrt/pull/16505/commits/c459a220d12bb352eb4f1fbd5caf32709ae7788a), but this is a cleanup commit that can be safely left out of the merge.

robimarko commented:
I tested this on 301w and it seems to work.

@Lanchon I assume this is now ready to be merged (I added the description to satisfy CI).

Lanchon commented:
@robimarko 

> I assume this is now ready to be merged

yes sir!

---

> (I added the description to satisfy CI)

sorry, didn't know it was a hard requirement!

robimarko commented:
Thanks for digging in and fixing these issues

Lanchon commented:
thank you!!

sppmasterspp commented:
> if you want to back up the partition as a binary image, and the partition is mostly empty: mount it, run `fstrim -v /path/to/device/mount/point/` to wipe unused space, unmount it, `gzip -c /dev/<emmc-dev>p9 > /tmp/p9.img.gz` to back it up, and then use scp to pull the image backup to your PC.
> 
> or you can just backup the files in an archive, which should be enough: `tar --acls --xattrs --selinux -cpvzf /tmp/p9.tgz /path/to/device/mount/point`. options --acls --xattrs and --selinux may be too much for busybox tar, you can leave them out then. or if you want an uncompressed backup (faster) you can use: `tar --acls --xattrs --selinux -cpvf /tmp/p9.tar /path/to/device/mount/point`.

@Lanchon Can you give here for reference the commands to restore the original files from the archives to partition 9 using gzip and tar.

Lanchon commented:
> @Lanchon Can you give here for reference the commands to restore the original files from the archives to partition 9 using gzip and tar.

`gzip -dc /tmp/p9.img.gz > /dev/<emmc-dev>p9` should work.

or else `gzip -dc /tmp/p9.img.gz | dd of=/dev/<emmc-dev>p9`

**but before you must make sure that partition p9 is unmounted!!!** use `mount` to check, `umount` to unmount.

---

restoring the tar backup requires formatting and mounting partition p9 beforehand. as backups go, it is less useful. but it allows easier access to the file contents from the backup file, if ever needed. the tar backup, strictly speaking, is redundant.
